File: DevelopArchitecture/DevelopArchitecture/architecture/baseVC/JBaseCollectionVC.m
//
//  JBaseCollectionVC.m
//  DevelopArchitecture
//
//  Created by Jeffrey on 15/10/24.
//  Copyright © 2015年 Jeffrey. All rights reserved.
//


#import "JBaseCollectionVC.h"

@interface JBaseCollectionVC ()<UICollectionViewDataSource,UICollectionViewDelegate,UICollectionViewDelegateFlowLayout>

@end

@implementation JBaseCollectionVC

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.view addSubview:self.m_collectionView];
}


-(UICollectionView*)m_collectionView{
    if(!_m_collectionView){
        _m_collectionView = kcreateCollectViewWithDelegate(self);
     }
    return _m_collectionView;
}

 
-(NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    if(self.m_isSearchTableViewList)return [self.m_searchDataArr count];
    else return [self.m_dataArr count];
    return 0;
}
-(NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView{
    return 1;
}
-(UICollectionViewCell*)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    
    JBaseCollectionViewCell *cell=[collectionView dequeueReusableCellWithReuseIdentifier:kCellIndentifier forIndexPath:indexPath];
    
    cell.backgroundColor = [UIColor colorWithRed:((255-10 * indexPath.row) / 255.0) green:((255-20 * indexPath.row)/255.0) blue:((30 * indexPath.row)/255.0) alpha:1.0f];
    return cell;
}
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath  {
    return CGSizeMake(KWidth/3-10, 100);
}
-(UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section  {
    return UIEdgeInsetsMake(0, 0, 0, 0);//分别为上、左、下、右
}
-(void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath{
    JBaseCollectionViewCell * cell = (JBaseCollectionViewCell *)[collectionView cellForItemAtIndexPath:indexPath];
    cell.backgroundColor=[UIColor whiteColor];
}
-(CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout referenceSizeForHeaderInSection:(NSInteger)section{
    return CGSizeMake(0, 0);//特别注意 横向滚动时，这两个值可能相反 导致第一页空格页
}
-(CGSize)collectionView:(UICollectionView *)collectionView layout:(nonnull UICollectionViewLayout *)collectionViewLayout referenceSizeForFooterInSection:(NSInteger)section{
    return CGSizeMake(0, 0);//特别注意 横向滚动时，这两个值可能相反 导致第一页空格页
}
- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath{
    
    NSString *reuseIdentifier;
    if ([kind isEqualToString: UICollectionElementKindSectionFooter ]){
        reuseIdentifier = @"footsection";
    }else{
        reuseIdentifier = @"headsection";
    }
    
    UICollectionReusableView *view =  [collectionView dequeueReusableSupplementaryViewOfKind :kind   withReuseIdentifier:reuseIdentifier   forIndexPath:indexPath];
    
    return view;
}
-(void)collectionView:(UICollectionView *)collectionView willDisplayCell:(UICollectionViewCell *)cell forItemAtIndexPath:(NSIndexPath *)indexPath{
    self.m_defaultImageView.hidden=YES;
}

- (BOOL)collectionView:(UICollectionView *)collectionView shouldSelectItemAtIndexPath:(NSIndexPath *)indexPath{
    return YES;
}
- (BOOL)collectionView:(UICollectionView *)collectionView shouldDeselectItemAtIndexPath:(NSIndexPath *)indexPath{
    return YES;
}
-(void )reloadData{
    [self.m_collectionView reloadData];
}

@end

File: DeveloperArchitecture/DeveloperArchitecture/architecture/baseVC/JBaseCollectionVC.swift
//
//  JBaseCollectionVC.swift
//  DeveloperArchitecture
//
//  Created by Jeffrey on 2020/10/30.
//  Copyright © 2020 Jeffery. All rights reserved.
//

import Foundation
import UIKit

class JBaseCollectionVC : JBaseTableViewVC ,UICollectionViewDelegate ,UICollectionViewDataSource ,UICollectionViewDelegateFlowLayout {
    @objc lazy var m_collectionView : UICollectionView = {
        let collectionView =  kcreateCollectViewWithDelegate(self)
        return collectionView
    }()
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        m_defaultImageView.isHidden = (m_collectionView.visibleCells.count > 0)
    }
    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.addSubview(m_collectionView)
    }
}

extension JBaseCollectionVC {
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }

    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.m_dataArr.count
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return CGSize(width: KWidth, height: 100)
    }
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: kCellIndentifier, for: indexPath)
        cell.backgroundColor = krandomColor
        return cell
    }
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let cell = collectionView.cellForItem(at: indexPath)
        cell?.backgroundColor = COLOR_WHITE_DARK
    }
}

extension JBaseCollectionVC {
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
        return UIEdgeInsets(top: 0, left: 0, bottom: 0, right: 0)//分别为上、左、下、右
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForHeaderInSection section: Int) -> CGSize {
        return CGSize(width: 0, height: 0) //特别注意 横向滚动时，这两个值可能相反 导致第一页空格页
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForFooterInSection section: Int) -> CGSize {
        return CGSize(width: 0, height: 0) //特别注意 横向滚动时，这两个值可能相反 导致第一页空格页
    }
    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
        var resultIdentifier = "headsection"
        if kind == UICollectionView.elementKindSectionFooter {
            resultIdentifier = "footsection"
        }
        let view = collectionView.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: resultIdentifier, for: indexPath)
        return view
    }
}

extension JBaseCollectionVC {
    func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
        self.m_defaultImageView.isHidden = true
    }
    func collectionView(_ collectionView: UICollectionView, shouldSelectItemAt indexPath: IndexPath) -> Bool {
        return true
    }
    func collectionView(_ collectionView: UICollectionView, shouldDeselectItemAt indexPath: IndexPath) -> Bool {
        return true
    }
}

func kcreateCollectViewWithDelegate(_ delegate : AnyObject) -> UICollectionView {
    let flowlayout = JCollectionViewFlowLayout()
    flowlayout.minimumLineSpacing = 2
    flowlayout.minimumInteritemSpacing = 0
    let collectionView = UICollectionView(frame: CGRect(x: 0, y: 0, width: KWidth, height: KHeight), collectionViewLayout: flowlayout)
    collectionView.showsVerticalScrollIndicator = false
    collectionView.showsHorizontalScrollIndicator = false
    collectionView.delegate = (delegate as? UICollectionViewDelegate)
    collectionView.dataSource = (delegate as? UICollectionViewDataSource)
    collectionView.allowsSelection = true
    collectionView.backgroundColor = UIColor.clear
    collectionView.register(JBaseCollectionViewCell.self, forCellWithReuseIdentifier: kCellIndentifier)
    collectionView.register(UICollectionReusableView.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "headsection")
    collectionView.register(UICollectionReusableView.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionFooter , withReuseIdentifier: "footsection")
    collectionView.alwaysBounceVertical = true
    if #available(iOS 11.0, *) {
        collectionView.contentInsetAdjustmentBehavior = .never
    }
    return collectionView
}
